Add TCP load balancing to deployment tutorials

jseldess opened the pull request:
- Add docs on using `cockroach gen haproxy`
- Add HAProxy load balancing to manual deployment docs
- Add GCE managed load balancing to GCE deployment docs
- Add AWS managed load balancing to AWS deployment docs
- Add Azure managed load balancing to Azure deployment docs
- Add Digital Ocean managed load balanced to DO deployment docs

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/docs/1213)
<!-- Reviewable:end -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/64798eb2e351b0f6b024a296af3223b508d5c1f9/

arjunravinarayan commented:
Only reviewed the files that I marked as read. LGTM, but do wait for @mberhault to weigh in as well.

---

Reviewed 3 of 5 files at r1.
Review status: 2 of 5 files reviewed at latest revision, 7 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-KfrtUSwBrEpyewI2BP6:b-ep134m) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
> ```Markdown
> 
> - Running at least 3 nodes to ensure survivability.
> - Running HAProxy on its own droplet. This ensures that when a CockroachDB node fails, HAProxy continues uninterrupted.
> ```

You should note that this setup means that the HAProxy node itself is a single-point of failure. In production you can run multiple HAProxy nodes. @mberhault can probably tell you better what people do in terms of defense-in-depth to protect against that.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrtp8r2r-CWkQGkAY9:-Kfrtp8r2r-CWkQGkAYA:b-ett6os) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L119)):*
> ```Markdown
> ## Step 5. Test your cluster
> 
> CockroachDB replicates and distributes data for you behind-the-scenes and uses a [Gossip protocol](https://en.wikipedia.org/wiki/Gossip_protocol) to guarantee that each node can locate data across the cluster. To test this, use the [built-in SQL client](use-the-built-in-sql-client.html) as follows:
> ```

"to guarantee that each node can locate" -> "to enable each node to locate". I wouldn't use the phrase gossip and guarantee in the same sentence.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfru_NPLIIYLYLE3kW3:-Kfru_NPLIIYLYLE3kW4:bol0j9q) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L198)):*
> ```Markdown
> 	~~~
> 
> 4. 	Run the `cockroach gen haproxy` command, specifying the interal IP address of any instance running a CockroachDB node:
> ```

interal -> internal

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrvNx-vNJt2qObBKBl:-KfrvNx-vNJt2qObBKBm:b-8dmq6f) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L203)):*
> ```Markdown
> 	$ cockroach gen haproxy \
> 	--host=<internal IP address of any node> \
> 	--port=26257 \
> ```

do note that if they roll the haproxy on the same node as a cockroach node they will have to use a different port. I can see this being a common mistake by someone cheaping out on droplets.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrv9msFsg-Y-pjekRr:-Kfrv9mt5yjiQYd3vU5l:b-zfahjt) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L210)):*
> ```Markdown
> 
> 	~~~ shell
> 	global
> ```

@mberhault does `cockroach gen haproxy` not specify a pid?

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrufN7OIVyMreldvFb:-KfrufN7OIVyMreldvFc:bgs0rhu) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L232)):*
> ```Markdown
> 	- For each node in the cluster, the `server` field specifies the interface that the node listens on, i.e., the address passed in the `--advertise-host` flag on node startup.
> 	- The `timout` fields specify timeout values that should be suitable for most deployments.
> 	- The `balance` field is set to the `roundrobin` balancing algorithm to ensure that traffic is spread as evenly as possible across nodes.
> ```

roundrobin won't do what you say it does in face of heterogenous loads. leastconn is better described as achieveing the goal of 'spreading traffic as evenly as possible', so I wouldn't describe roundrobin that way. roundrobin just rotates across all the listed servers in order. It's good enough, so to speak, and we can perhaps mention that if they have heterogenous loads (say some long lived connections and lots of short lived connections) then leastconn might be a better setting for them.

---

*[load-balancing.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtCFNRkmJsMIbnJjm:-KfrtCFNRkmJsMIbnJjn:b-bybubh) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/load-balancing.md#L9)):*
> ```Markdown
> CockroachDB replicates and distributes data for you behind-the-scenes and uses a [Gossip protocol](https://en.wikipedia.org/wiki/Gossip_protocol) to guarantee that each node can locate data across the cluster. As a result, each node is an equally suitable SQL gateway; clients can connect to any node and CockroachDB will route their requests automatically.
> 
> However, it's important to note that a gateway node does a lot of work. In addition to routing queries, it receives data from other nodes, performs its own computations, and returns results to the client. Sending all client requests to a single node will therefore result in poor performance (queries per second). Instead, it's recommended to balance client workloads across nodes for more efficient and performant processing.
> ```

can rephrase poor performance -> suboptimal performance. Poor is perhaps a little too strong.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-:-KfrviYht_RAxDwZAXpZ:b-p3t76a)*
<!-- Sent from Reviewable.io -->

mberhault commented:
Review status: 2 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[deploy-cockroachdb-on-aws-insecure.md, line 251 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfuGykQouSnYxtyLcsJ:-KfuGykQouSnYxtyLcsK:b-91ddzh) ([raw file](https://github.com/cockroachdb/docs/blob/64798eb2e351b0f6b024a296af3223b508d5c1f9/deploy-cockroachdb-on-aws-insecure.md#L251)):*
> ```Markdown
> 	~~~
> 
> 	The file is preset with the minimal configurations needed to work with your running cluster:
> ```

we should mention that the `bind : 26257` line specifies which port haproxy listens on. This is the port clients will connect to and needs to be allowed by the firewall.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-KfuHxCqu_qAk07pY_tk:b-n0vt0o) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

You should note that this setup means that the HAProxy node itself is a single-point of failure. In production you can run multiple HAProxy nodes. @mberhault can probably tell you better what people do in terms of defense-in-depth to protect against that.
</blockquote></details>

people may run multiple load balancers (or rely on cloud-run ones) and use DNS to pick one. I don't think we need to get into that discussion though, we just want to show how to run haproxy, not give non-cockroach production recommendations.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrvNx-vNJt2qObBKBl:-KfuIMX4sVYw094OG-9z:bms14oz) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L203)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

do note that if they roll the haproxy on the same node as a cockroach node they will have to use a different port. I can see this being a common mistake by someone cheaping out on droplets.
</blockquote></details>

although this is not the haproxy port, this is the port used to talk to cockroach.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrv9msFsg-Y-pjekRr:-KfuHJ34tFIeEHVsvLAU:b-kmd5oi) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L210)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

@mberhault does `cockroach gen haproxy` not specify a pid?
</blockquote></details>

you mean a pid file? we don't specify that.
the docs (the ones here, and the help message on the haproxy command) do mention that this is a minimal situation. I did make an exception for the `maxconn` option, but even that we could leave out.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrufN7OIVyMreldvFb:-KfuHOijdQ_vMtr7Awou:b-407d0y) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L232)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

roundrobin won't do what you say it does in face of heterogenous loads. leastconn is better described as achieveing the goal of 'spreading traffic as evenly as possible', so I wouldn't describe roundrobin that way. roundrobin just rotates across all the listed servers in order. It's good enough, so to speak, and we can perhaps mention that if they have heterogenous loads (say some long lived connections and lots of short lived connections) then leastconn might be a better setting for them.
</blockquote></details>

fair enough, but we don't want to go into a discussion of what the right load-balancing method is. RR is one of the simplest and most easily understood. However, this could be rephrased to:
`... to ensure that traffic is spread across nodes`.  "evenly" gets us into trouble.

---

*[load-balancing.md, line 2 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfuHlobL3MlY3GGnxFA:-KfuHlobL3MlY3GGnxFB:b-1gs9pf) ([raw file](https://github.com/cockroachdb/docs/blob/64798eb2e351b0f6b024a296af3223b508d5c1f9/load-balancing.md#L2)):*
> ```Markdown
> ---
> title: Load Balancing
> ```

where is this file used? it doesn't seem to be included by anything else. Is it hugo magic?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

jseldess commented:
Review status: 2 of 5 files reviewed at latest revision, 9 unresolved discussions.

---

*[deploy-cockroachdb-on-aws-insecure.md, line 251 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfuGykQouSnYxtyLcsJ:-KfvR8DPI4UAmE3nUWk3:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/64798eb2e351b0f6b024a296af3223b508d5c1f9/deploy-cockroachdb-on-aws-insecure.md#L251)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

we should mention that the `bind : 26257` line specifies which port haproxy listens on. This is the port clients will connect to and needs to be allowed by the firewall.
</blockquote></details>

Done.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-Kfv_kyUqwrckkjLX0Yo:b-4uyyiu) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

people may run multiple load balancers (or rely on cloud-run ones) and use DNS to pick one. I don't think we need to get into that discussion though, we just want to show how to run haproxy, not give non-cockroach production recommendations.
</blockquote></details>

I added a note in the `Create Droplets` step, since that's where we first recommend a separate Droplet for HAProxy.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 119 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrtp8r2r-CWkQGkAY9:-KfvRSH3WeOsAOJc3vwa:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L119)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

"to guarantee that each node can locate" -> "to enable each node to locate". I wouldn't use the phrase gossip and guarantee in the same sentence.
</blockquote></details>

Done.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 198 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfru_NPLIIYLYLE3kW3:-KfvRZdJQ9V0AIwt_xTN:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L198)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

interal -> internal
</blockquote></details>

Done.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 203 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrvNx-vNJt2qObBKBl:-KfvTXayNmtctMJBjawX:b-k63mbt) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L203)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

although this is not the haproxy port, this is the port used to talk to cockroach.
</blockquote></details>

@arjunravinarayan, I've added that node in the description of the `bind` field in the haproxy config.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrv9msFsg-Y-pjekRr:-KfvVQ7mfutugo4XNv0j:b-txadvc) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L210)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

you mean a pid file? we don't specify that.
the docs (the ones here, and the help message on the haproxy command) do mention that this is a minimal situation. I did make an exception for the `maxconn` option, but even that we could leave out.
</blockquote></details>

Since we don't specify this, some default gets used, right? Is that important to call out? Also, what's this file for?

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrufN7OIVyMreldvFb:-KfvWyP9JsoAoOf6ao6K:b-un7buf) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L232)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

fair enough, but we don't want to go into a discussion of what the right load-balancing method is. RR is one of the simplest and most easily understood. However, this could be rephrased to:
`... to ensure that traffic is spread across nodes`.  "evenly" gets us into trouble.
</blockquote></details>

Tweaked/expanded this language a bit. Seems fine to add a few hints. Marc, please let me know if you dislike anything I've added.

---

*[load-balancing.md, line 9 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtCFNRkmJsMIbnJjm:-KfvaH6nu18FYYxnyKIO:b-896fix) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/load-balancing.md#L9)):*
<details><summary><i>Previously, arjunravinarayan (Arjun Narayan) wrote…</i></summary><blockquote>

can rephrase poor performance -> suboptimal performance. Poor is perhaps a little too strong.
</blockquote></details>

Done.

---

*[load-balancing.md, line 2 at r2](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfuHlobL3MlY3GGnxFA:-Kfva1XTznm4w57ew4gG:b2uemw) ([raw file](https://github.com/cockroachdb/docs/blob/64798eb2e351b0f6b024a296af3223b508d5c1f9/load-balancing.md#L2)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

where is this file used? it doesn't seem to be included by anything else. Is it hugo magic?
</blockquote></details>

I haven't decided if/how to use this page yet. My initial thought is that it'd be nice to have a single page on "load balancing" that links to relevant tutorials. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f1344907ae645159e89742e7e25c58b03b6ee35a/

mberhault commented:
Review status: 1 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrv9msFsg-Y-pjekRr:-Kfve0m6QMraZ3NfxTr7:b-3qxrcv) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L210)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Since we don't specify this, some default gets used, right? Is that important to call out? Also, what's this file for?
</blockquote></details>

sure, but this is just one of many options in the config file, I don't think we should start explaining them. We should have a link to the haproxy config docs though.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrufN7OIVyMreldvFb:-Kfvfj3TPoE4ogMn8XBP:bwp76ui) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L232)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

Tweaked/expanded this language a bit. Seems fine to add a few hints. Marc, please let me know if you dislike anything I've added.
</blockquote></details>

I would keep it simple, haproxy has too many balancing algorithms to get into, and even leastconn is not particularly helpful.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

mberhault commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Review status: 1 of 5 files reviewed at latest revision, 8 unresolved discussions.

---



*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-:-KfvgWebAKVRT8xlpEgh:bnfp4nl)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 1 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-Kfvn6MyE2AWhV74XWiU:b-ijyj8q) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, jseldess wrote…</i></summary><blockquote>

I added a note in the `Create Droplets` step, since that's where we first recommend a separate Droplet for HAProxy.
</blockquote></details>

I think we do want to talk about running multiple haproxies, though. This bullet implies that running a single haproxy is somehow more reliable than pointing your clients to a single cockroach node. 

We should make it clear that there are two reasons to use a load balancer: for performance (to spread the load across all the cockroach nodes when it is impractical for you to configure your clients with all the individual addresses), and for reliability (in which case you must use more than one load balancer to have a meaningful improvement).

In the performance case, it's still a good idea to run haproxy on its own machine, but it's to avoid interference between haproxy and the cockroach process, instead of for reliability. 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

sploiselle commented:
Review status: 1 of 5 files reviewed at latest revision, 8 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-KfvpKkM2ciwyKy6sj7H:b-r6tan1) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, bdarnell (Ben Darnell) wrote…</i></summary><blockquote>

I think we do want to talk about running multiple haproxies, though. This bullet implies that running a single haproxy is somehow more reliable than pointing your clients to a single cockroach node. 

We should make it clear that there are two reasons to use a load balancer: for performance (to spread the load across all the cockroach nodes when it is impractical for you to configure your clients with all the individual addresses), and for reliability (in which case you must use more than one load balancer to have a meaningful improvement).

In the performance case, it's still a good idea to run haproxy on its own machine, but it's to avoid interference between haproxy and the cockroach process, instead of for reliability. 
</blockquote></details>

@bdarnell I'm using haproxy for curriculum I'm developing around survivability and am planning on leveraging these docs.

Is there a procedure you recommend for using multiple haproxy nodes? Does [this Digital Ocean guide](https://www.digitalocean.com/community/tutorials/how-to-create-a-high-availability-haproxy-setup-with-corosync-pacemaker-and-floating-ips-on-ubuntu-14-04) seem reasonable?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

bdarnell commented:
Review status: 1 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-KfvqZXrKCz9wy9-XmxK:b-9acgsp) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

@bdarnell I'm using haproxy for curriculum I'm developing around survivability and am planning on leveraging these docs.

Is there a procedure you recommend for using multiple haproxy nodes? Does [this Digital Ocean guide](https://www.digitalocean.com/community/tutorials/how-to-create-a-high-availability-haproxy-setup-with-corosync-pacemaker-and-floating-ips-on-ubuntu-14-04) seem reasonable?
</blockquote></details>

Yeah, that guide looks fine, but I think in most cases we should be recommending the cloud-specific load balancers (https://www.digitalocean.com/community/tutorials/an-introduction-to-digitalocean-load-balancers for digital ocean, ELBs for amazon, etc)

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213)*
<!-- Sent from Reviewable.io -->

arjunravinarayan commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

---

Reviewed 2 of 3 files at r3.
Review status: 3 of 5 files reviewed at latest revision, 4 unresolved discussions.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 52 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrtUSvhxliBY39a8Ui:-KfvqZq5GI44mNClA-NI:b-1jwb4c) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L52)):*
<details><summary><i>Previously, sploiselle (Sean Loiselle) wrote…</i></summary><blockquote>

@bdarnell I'm using haproxy for curriculum I'm developing around survivability and am planning on leveraging these docs.

Is there a procedure you recommend for using multiple haproxy nodes? Does [this Digital Ocean guide](https://www.digitalocean.com/community/tutorials/how-to-create-a-high-availability-haproxy-setup-with-corosync-pacemaker-and-floating-ips-on-ubuntu-14-04) seem reasonable?
</blockquote></details>

I'm with @bdarnell here. We should at the very least make it clear that a single load balancing node _is_ a single point of failure. Beyond that, I do like the addition of mentioning a paid load balancing service that comes with its own redundancy and failover guarantees.
@sploiselle that guide seems very reasonable.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 210 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-Kfrv9msFsg-Y-pjekRr:-KfvqxMh7zlShx3yz5C5:busvl7n) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L210)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

sure, but this is just one of many options in the config file, I don't think we should start explaining them. We should have a link to the haproxy config docs though.
</blockquote></details>

Marc's guidelines work for me.

---

*[deploy-cockroachdb-on-digital-ocean-insecure.md, line 232 at r1](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-KfrufN7OIVyMreldvFb:-Kfvr4moBy07AaicQFNZ:b-imwqk1) ([raw file](https://github.com/cockroachdb/docs/blob/99f9fec5381b2b5c4ef78ff59a2d416daa18e723/deploy-cockroachdb-on-digital-ocean-insecure.md#L232)):*
<details><summary><i>Previously, mberhault (marc) wrote…</i></summary><blockquote>

I would keep it simple, haproxy has too many balancing algorithms to get into, and even leastconn is not particularly helpful.
</blockquote></details>

Marc's suggestion is fine by me.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-:-KfvrC6XryLvIYvk949n:bnfp4nl)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/9a7af06aa8f6642f32a637745c058ecaabe172a8/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/125b9acc2b30f7d95e584425464e019085aef765/

jseldess commented:
@mberhault, @arjunravinarayan, @bdarnell, can someone please take one last look? I revised the Manual Deployment docs to feature HAProxy, and used the platform-specific load balancing service in our cloud deployment docs. Also added docs on the `cockroach gen haxproxy` command. 

---







*Comments from [Reviewable](https://reviewable.io:443/reviews/cockroachdb/docs/1213#-:-KgR14MtxcpUzmgY8PTf:bdvwi28)*
<!-- Sent from Reviewable.io -->

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/f07e3951b7b980c234e1ab0000fa4c9287fd07ea/

mberhault commented on deploy-cockroachdb-on-aws.md:
```diff
@@ -69,22 +69,37 @@ You can create these rules using [Security Groups' Inbound Rules](http://docs.aw
+
+1. 	[Add AWS load balancing](http://docs.aws.amazon.com/AWSEC2/latest/UserGuide/ec2-increase-availability.html). Be sure to:
+	- Set forwarding rules to route TCP traffic from the load balancer's port **26257** to port **26257** on the node Droplets.
+	- Configure health checks to use HTTP port **8080** and path `/health`.
```
this section is for secure mode. The AWS load balancer must be told to use https

jseldess commented:
Thanks, @mberhault. It's very time-consuming, but I'll try to test this on more platforms. 

Just tested secure on Digital Ocean. No additional settings are required for secure there. Even without "Redirect HTTP to HTTPs" turned on. 

Moving on to GCE now.

jseldess commented:
Also, @arjunravinarayan offered to test out the instructions on AWS.

jseldess commented:
@mberhault, I added specific instructions for the GCE tutorials. Steps work fine for both secure and insecure. Please let me know what you think. I'll do Azure next.

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/c9e3991b4c1c45ab20d1784a20be720beb5f2805/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/a2d53db920cbe5c42a2a86100e46647a5db49e43/

cockroach-teamcity commented:
http://cockroach-docs-review.s3-website-us-east-1.amazonaws.com/10f872ec4ec35ebf1ad56aec872156340741ff22/

jseldess commented:
@mberhault, @arjunravinarayan, given that the manual deployment, digital ocean, and gce docs are ok at this point, I'm going to merge this to include some of these updates in today's release notes. I'll open a follow-up PR to improve the aws and azure instructions.
